tool/site/client-react/src/components/longRunningTask.tsx

```diff
@@ -0,0 +1,75 @@
+import * as React from "react"
+import {JSXElementConstructor, useContext, useEffect, useState} from "react"
+import {Button, ErrorMessage, Label} from "nhsuk-react-components"
+import ButtonList from "./buttonList"
+import {AppContext} from "../index"
+
+interface LongRunningTaskProps<T> {
+  task: () => Promise<T>
+  loadingMessage: string
+  children: JSXElementConstructor<T>
+  back?: string | (() => void)
+}
+
+const LongRunningTask = <T extends unknown>({
+  task,
+  loadingMessage,
+  children,
+  back
+}: LongRunningTaskProps<T>): React.ReactElement => {
+  const {baseUrl} = useContext(AppContext)
+  const [loading, setLoading] = useState<boolean>(true)
+  const [errorMessage, setErrorMessage] = useState<string>()
+  const [result, setResult] = useState<T>()
+
+  if (!back) {
+    back = baseUrl
+  }
+
+  useEffect(() => {
+    async function performTask() {
+      setLoading(true)
+      try {
+        const loadResult = await task()
+        setResult(loadResult)
+      } catch (e) {
+        console.log(e)
+        setErrorMessage(e.message)
+      } finally {
+        setLoading(false)
+      }
+    }
+
+    if (!result) {
+      performTask()
```

Comment:
Don't _need_ to, but it might make sense to. Will try it and see if anything breaks

Reply:
Quick google suggests changing it to
```
useEffect(() => {
    if (!result) {
      (async () => {
        setLoading(true)
        try {
          const loadResult = await task()
          setResult(loadResult)
        } catch (e) {
          console.log(e)
          setErrorMessage(e.message)
        } finally {
          setLoading(false)
        }
      })()
    }
  }, [result, task])
```
Which still doesn't await the result of the IIFE, but looks a bit neater to me

Reply:
What doesn't work is this
```
  useEffect(async () => {
    if (!result) {
      setLoading(true)
      try {
        const loadResult = await task()
        setResult(loadResult)
      } catch (e) {
        console.log(e)
        setErrorMessage(e.message)
      } finally {
        setLoading(false)
      }
    }
  }, [result, task])
```
Because you're not allowed to return a promise from the useEffect callback

Reply:
👍